org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenImpl.java

```diff
@@ -714,56 +713,104 @@ private Properties loadLastUpdated(ArtifactRepository localRepository, Artifact
   }
 
   private File getLastUpdatedFile(ArtifactRepository localRepository, Artifact artifact) {
-    return new File(localRepository.getBasedir(), basePathOf(artifact) + "/m2e-lastUpdated.properties");
+    return Path.of(localRepository.getBasedir(), //
+        artifact.getGroupId().replace('.', '/'), artifact.getArtifactId(), artifact.getBaseVersion(),
+        "m2e-lastUpdated.properties").toFile();
+  }
+
+  @Override
+  public ConfigurationElement getMojoConfiguration(MavenProject project, MojoExecution execution,
+      IProgressMonitor monitor) throws CoreException {
+    return getExecutionContext().execute(project, (context, pm) -> {
+      try {
+        MavenSession session = context.getSession();
+        MojoDescriptor mojo = execution.getMojoDescriptor();
+        ClassRealm pluginRealm = lookup(BuildPluginManager.class).getPluginRealm(session, mojo.getPluginDescriptor());
+        PluginParameterExpressionEvaluator evaluator = new PluginParameterExpressionEvaluator(session, execution);
+        ValueComputer valueComputer = new ValueComputer(converterLookup, mojo, pluginRealm, evaluator);
+
+        Xpp3Dom dom = execution.getConfiguration();
+        PlexusConfiguration configuration = dom != null ? new XmlPlexusConfiguration(dom) : null;
+        return new ConfigurationElementImpl(configuration, "", valueComputer);
+      } catch(Exception e) {
+        throw new CoreException(Status.error(
+            NLS.bind("Could not get the configuration for for plugin execution {0}", execution.getExecutionId()), e));
+      }
+    }, monitor);
   }
 
-  private static final char PATH_SEPARATOR = '/';
+  private record ValueComputer(ConverterLookup lookup, MojoDescriptor mojo, ClassLoader pluginRealm,
+      ExpressionEvaluator evaluator) {
 
-  private static final char GROUP_SEPARATOR = '.';
+    private <T> T convertConfigurationValueToType(PlexusConfiguration configuration, Class<T> clazz)
+        throws ComponentConfigurationException {
 
-  private String basePathOf(Artifact artifact) {
-    return formatAsDirectory(artifact.getGroupId()) + PATH_SEPARATOR + artifact.getArtifactId() + PATH_SEPARATOR
-        + artifact.getBaseVersion() + PATH_SEPARATOR;
+      ConfigurationConverter typeConverter = lookup.lookupConverterForType(clazz);
+      Object value = typeConverter.fromConfiguration(lookup, configuration, clazz, mojo.getImplementationClass(),
+          pluginRealm, evaluator, null);
+      return clazz.cast(value);
+    }
   }
 
-  private String formatAsDirectory(String directory) {
-    return directory.replace(GROUP_SEPARATOR, PATH_SEPARATOR);
+  private class ConfigurationElementImpl implements ConfigurationElement {
```

Comment:
I would put this and related classes to a dedicated file.

Reply:
I have no strong opinion about that.
But the `MavenImpl` class is already quite large and moving the configuration related classes into a dedicated class makes it simpler to overview that story.